server/src/main/java/org/apache/druid/server/coordination/ChangeRequestHistory.java

```diff
@@ -74,11 +76,24 @@ public ChangeRequestHistory(int maxSize)
     this.singleThreadedExecutor = Execs.singleThreaded("SegmentChangeRequestHistory");
   }
 
+  public void stop()
+  {
+    singleThreadedExecutor.shutdownNow();
+    final LinkedHashSet<CustomSettableFuture<?>> futures = new LinkedHashSet<>(waitingFutures.keySet());
+    waitingFutures.clear();
+    for (CustomSettableFuture<?> theFuture : futures) {
+      theFuture.setException(new IllegalStateException("Server is shutting down."));
+    }
+  }
+
   /**
    * Add batch of segment changes update.
    */
   public synchronized void addChangeRequests(List<T> requests)
   {
+    if (singleThreadedExecutor.isShutdown()) {
```

Comment:
should this also throw an exception - since otherwise the caller might never know that the operation is illegal

Reply:
does it matter since the server is shutting down?

Reply:
I thought this method could end up being called from a `HttpLoadQueuePeon` API request (`/changeRequests`), so would be better if we can return the correct status/error to it as soon as possible. wdyt?

Reply:
it can, but I guess my thoughts are that it doesn't matter since the `HttpLoadQueuePeon` is stopped by the server disappearing from discovery, so it seemed better to silently ignore because its going down anyway and to avoid any extra log spam and not have potentially multiple shutdown paths on the `HttpLoadQueuePeon` side (same reason I opted to go with failing the future with exception instead of a successful "im shutting dowm" response).

Reply:
Thanks for the clarification - letting all shutdowns being managed by the discovery makes sense to me!